src/metabase/util/ssh.clj

```diff
@@ -1,32 +1,63 @@
    on the returned connection object."
-  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass host port]}]
-  (let [connection (doto ^com.jcraft.jsch.Session (.getSession (new com.jcraft.jsch.JSch)
-                                                               ^String tunnel-user
-                                                               ^String tunnel-host
-                                                               tunnel-port)
-                     (.setPassword ^String tunnel-pass)
-                     (.setConfig "StrictHostKeyChecking" "no")
-                     (.connect default-ssh-timeout)
-                     (.setPortForwardingL 0 host port))
-        input-port (some-> (.getPortForwardingL connection)
-                           first
-                           (str/split #":")
-                           first
-                           (Integer/parseInt))]
-    (assert (number? input-port))
-    (log/info (u/format-color 'cyan "creating ssh tunnel %s@%s:%s -L %s:%s:%s" tunnel-user tunnel-host tunnel-port input-port host port))
-    [connection input-port]))
+  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass tunnel-private-key tunnel-private-key-passphrase host port]}]
+  (let [conn-future (.connect ^SshClient client ^String tunnel-user ^String tunnel-host ^Integer tunnel-port)
+        conn-status (.verify ^ConnectFuture conn-future default-ssh-timeout)
+        session (.getSession ^SessionHolder conn-status)
+
+        _ (when tunnel-pass
+            (.addPasswordIdentity ^ClientSession session tunnel-pass))
+        _ (when tunnel-private-key
+            (let [resource-key (proxy [AbstractIoResource] [(class "key") "key"])
+                  password-provider (proxy [FilePasswordProvider] []
+                                      (getPassword [_ _ _]
+                                        tunnel-private-key-passphrase)
+                                      (handleDecodeAttemptResult [_ _ _ _ _]
+                                        FilePasswordProvider$ResourceDecodeResult/TERMINATE))
+
+                  ids (SecurityUtils/loadKeyPairIdentities ^ClientSession session
+                                                           resource-key
+                                                           (ByteArrayInputStream. (.getBytes ^String tunnel-private-key "UTF-8"))
+                                                           password-provider)
+
+                  keypair (GenericUtils/head ids)]
+              (.addPublicKeyIdentity ^ClientSession session keypair)))
+
+        auth-future (.auth ^ClientSession session)
+        _ (.verify ^AuthFuture auth-future default-ssh-timeout)
+
+        tracker (.createLocalPortForwardingTracker ^ClientSession session
+                                                   (SshdSocketAddress. "" 0)
+                                                   (SshdSocketAddress. host port))
+
```

Comment:
Not saying this is blocking merge, but for next time. https://guide.clojure.style/#no-blank-lines-within-def-forms (test definitions are an exception to this)